backend/EkoFunkcje/EkoFunkcje/Auth/IAuth.cs

```diff
@@ -0,0 +1,110 @@
+    {
+        bool IsAuthorized(HttpRequest request, string functionName);
+        Role GetUserRole(HttpRequest request);
+    }
+
+    public class Auth : IAuth
+    {
+        private readonly IDictionary<string, Role> _roles;
+        private IDictionary<string, Role> _userRoles;
+        private readonly System.Timers.Timer _timer;
+        private CloudTable _cloudTable;
+        private readonly string _connectionString;
+        private readonly ILogger<Auth> _logger;
+
+        private Auth(IDictionary<string, Role> roles, string storageConnectionString, ILogger<Auth> logger)
+        {
+            _logger = logger;
+            _roles = roles ?? new Dictionary<string, Role>();
+            _connectionString = storageConnectionString;
+            _timer = new System.Timers.Timer(3600000)
+            {
+                AutoReset = true
+            };
+            _timer.Elapsed += TimerElapsed;
+            _timer.Start();
+        }
+
+        public static async Task<Auth> InitializeAuthAsync(IDictionary<string, Role> roles, string storageConnectionString,
+            ILogger<Auth> logger)
+        {
+            var auth = new Auth(roles, storageConnectionString, logger);
+            await auth.InitializeDatabaseConnection();
+            auth._userRoles = await auth.LoadUserRoles();
+            return auth;
+        }
+
+        private async Task InitializeDatabaseConnection()
+        {
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(_connectionString);
+            CloudTableClient tableClient = storageAccount.CreateCloudTableClient();
+            _cloudTable = tableClient.GetTableReference(Config.UserRolesTableName);
+            await _cloudTable.CreateIfNotExistsAsync();
+        }
+
+        private void TimerElapsed(object sender, ElapsedEventArgs e)
+        {
+            var userRoles = LoadUserRoles().GetAwaiter().GetResult();
+            Interlocked.Exchange(ref _userRoles, userRoles);
+        }
+
+        private async Task<IDictionary<string, Role>> LoadUserRoles()
+        {
+            TableContinuationToken token = null;
+            var entities = new List<UserRole>();
+            do
+            {
+                var queryResult = await _cloudTable.ExecuteQuerySegmentedAsync(new TableQuery<UserRole>(), token);
+                entities.AddRange(queryResult.Results);
+                token = queryResult.ContinuationToken;
+            } while (token != null);
+            return entities.ToImmutableDictionary(e => e.UserId, e => e.Role);
+        }
+
+        public bool IsAuthorized(HttpRequest request, string functionName)
+        {
+            if (!request.IsAuthenticated())
+                return false;
+            if (!_roles.ContainsKey(functionName))
+                return true;
+            var roleNeeded = _roles[functionName];
+            var userName = request.HttpContext.User.Identity.Name;
+            if (!_userRoles.ContainsKey(userName))
```

Comment:
1. Wydaje mi się, że trochę łatwiej byłoby pozbyć się w ogóle pola _roles i podawać rolę jako parametr do metody podczas jej wywołania. Efekt ten sam a nie ma pośredniej stuktury danych :) Cala klasa by się wtedy uprościła. A użycie metody też by było czytelniejsze na poziomie poszczególnych funkcji. Co o tym myślisz?
```
public bool IsAuthorized(HttpRequest request, Role role = Role.Empty)
```

2. Przy obecnej implementacji, logika tej funkcji zachowuje sie inaczej niz GetUserRole.
Co myślisz o tym, żeby oprzeć ją całkowicie na GetUserRole, coś jak przykład poniżej?
```
public bool IsAuthorized(HttpRequest request, Role role = Role.Empty) {
      return GetUserRole(request) >= role;
}
```

Reply:
Również dobry pomysł. W sumie zrobiłem ta strukture zeby pozniej ja udostepniac dla frontu. Tj zrobic api ktore zwraca wartosci tej struktury czyli jaki widok - jakie uprawnienia wymagane. Wtedy będzie można to na froncie pokazywać/ukrywać. Ale możemy to też śmialo inaczej zrobić
Co myślisz?

Reply:
@CorneeSean up